modules/adagioRtdProvider.js

```diff
@@ -0,0 +1,680 @@
+  const w = (canAccessWindowTop()) ? getWindowTop() : getWindowSelf();
+
+  w.ADAGIO = w.ADAGIO || {};
+  w.ADAGIO.pageviewId = w.ADAGIO.pageviewId || generateUUID();
+  w.ADAGIO.adUnits = w.ADAGIO.adUnits || {};
+  w.ADAGIO.pbjsAdUnits = w.ADAGIO.pbjsAdUnits || [];
+  w.ADAGIO.queue = w.ADAGIO.queue || [];
+  w.ADAGIO.windows = w.ADAGIO.windows || [];
+
+  return w.ADAGIO;
+})();
+
+/**
+ * Store the sampling data.
+ * This data is used to determine if beacons should be sent to adagio.
+ * The sampling data
+ */
+const _SESSION = (function() {
+  /**
+   * @type {SessionData}
+   */
+  const data = {
+    session: {}
+  };
+
+  return {
+    init: () => {
+      // helper function to determine if the session is new.
+      const isNewSession = (lastActivity) => {
+        const now = Date.now();
+        return (!isNumber(lastActivity) || (now - lastActivity) > SESS_DURATION)
+      }
+
+      storage.getDataFromLocalStorage('adagio', (storageValue) => {
+        // session can be an empty object
+        const { rnd, new: isNew, vwSmplg, vwSmplgNxt, lastActivityTime } = _internal.getSessionFromLocalStorage(storageValue);
+
+        data.session = {
+          rnd,
+          new: isNew || false, // legacy: `new` was used but the choosen name is not good.
+          // Don't use values if they are not defined.
+          ...(vwSmplg !== undefined && { vwSmplg }),
+          ...(vwSmplgNxt !== undefined && { vwSmplgNxt }),
+          ...(lastActivityTime !== undefined && { lastActivityTime })
+        }
+
+        if (isNewSession(lastActivityTime)) {
+          data.session.new = true;
+          data.session.rnd = Math.random();
+        }
+
+        _internal.getAdagioNs().queue.push({
+          action: 'session',
+          ts: Date.now(),
+          data: {
+            session: {
+              ...data.session
+            }
+          }
+        });
+      });
+    },
+    get: function() {
+      return data.session;
+    }
+  }
+})();
+
+const _FEATURES = (function() {
+  /**
+   * @type {Features}
+   */
+  const features = {
+    initialized: false,
+    data: {},
+  };
+
+  return {
+    // reset is used for testing purpose
+    reset: function() {
+      features.initialized = false;
+      features.data = {};
+    },
+    get: function() {
+      if (!features.initialized) {
+        features.data = {
+          page_dimensions: getPageDimensions().toString(),
+          viewport_dimensions: getViewPortDimensions().toString(),
+          user_timestamp: getTimestampUTC().toString(),
+          dom_loading: getDomLoadingDuration().toString(),
+        }
+        features.initialized = true;
+      }
+
+      return { ...features.data };
+    }
+  }
+})();
+
+export const _internal = {
+  getAdagioNs: function() {
+    return _ADAGIO;
+  },
+
+  getSession: function() {
+    return _SESSION;
+  },
+
+  getFeatures: function() {
+    return _FEATURES;
+  },
+
+  /**
+   * Ensure that the bidder is Adagio.
```

Comment:
RTD modules cannot favor a particular hardcoded bidder. They can favor bidders if the list is taken from publisher configuration; but they cannot 'compute features only for the adagio bidder'

Reply:
Ok, I'll do the changes, it was already planed anyway. 
Beside this required change, is there anything that is not compliant?

Reply:
no other than this looks good!